astropy/modeling/rotations.py

```diff
@@ -27,70 +27,114 @@
 
-    def _create_matrix(self, phi, theta, psi, axes_order):
-        matrices = []
-        for angle, axis in zip([phi, theta, psi], axes_order):
-            if isinstance(angle, u.Quantity):
-                angle = angle.value
-            angle = angle.item()
-            matrices.append(rotation_matrix(angle, axis, unit=u.rad))
-        result = matrix_product(*matrices[::-1])
-        return result
+    angles = Parameter(getter=np.rad2deg, setter=np.deg2rad)
 
-    @staticmethod
-    def spherical2cartesian(alpha, delta):
-        alpha = np.deg2rad(alpha)
-        delta = np.deg2rad(delta)
-        x = np.cos(alpha) * np.cos(delta)
-        y = np.cos(delta) * np.sin(alpha)
-        z = np.sin(delta)
-        return np.array([x, y, z])
+    def __init__(self, angles, axes_order, name=None):
+        self.axes = ['x', 'y', 'z']
+        unrecognized = set(axes_order).difference(self.axes)
+        if unrecognized:
+            raise ValueError("Unrecognized axis label {0}; "
+                             "should be one of {1} ".format(unrecognized,
+                                                            self.axes))
+        self.axes_order = axes_order
+        if len(angles) != len(axes_order):
+            raise ValueError("The number of angles {0} should match the number \
+                              of axes {1}.".format(len(angles),
+                                                   len(axes_order)))
+        super(Rotation3D, self).__init__(angles, name=name)
+        self.inputs = ('x', 'y', 'z')
+        self.outputs = ('x', 'y', 'z')
 
-    @staticmethod
-    def cartesian2spherical(x, y, z):
-        h = np.hypot(x, y)
-        alpha = np.rad2deg(np.arctan2(y, x))
-        delta = np.rad2deg(np.arctan2(z, h))
-        return alpha, delta
+    @property
+    def inverse(self):
+        """Inverse rotation."""
+        angles = self.angles.value[::-1] * -1
+        return self.__class__(angles, axes_order=self.axes_order[::-1])
 
-    @deprecated(2.0)
-    @staticmethod
-    def rotation_matrix_from_angle(angle):
+    def evaluate(self, x, y, z, angles):
         """
-        Clockwise rotation matrix.
-
-        Parameters
-        ----------
-        angle : float
-            Rotation angle in radians.
+        Apply the rotation to a set of 3D Cartesian coordinates.
         """
-        return np.array([[math.cos(angle), math.sin(angle)],
-                         [-math.sin(angle), math.cos(angle)]])
+        if x.shape != y.shape != z.shape:
+            raise ValueError("Expected input arrays to have the same shape")
+        # Note: If the original shape was () (an array scalar) convert to a
+        # 1-element 1-D array on output for consistency with most other models
+        orig_shape = x.shape or (1,)
+        inarr = np.array([x.flatten(), y.flatten(), z.flatten()])
+        result = np.dot(_create_matrix(angles[0], self.axes_order), inarr)
+        x, y, z = result[0], result[1], result[2]
```

Comment:
Is this line necessary? Why not `x, y, z = np.dot(...)` directly?

Reply:
fixed